docs/doc_sources/beginners_guides/tensor_intro.rst

```diff
@@ -0,0 +1,287 @@
+
+.. code-block:: python
+    :caption: Example: ``asarray`` may migrate multiple arrays
+
+    from dpctl import tensor
+    import numpy
+
+    x_cpu = tensor.ones((10, 10), device="cpu")
+    x_gpu = tensor.zeros((10, 10), device="opencl:gpu")
+    x_np = numpy.random.randn(10, 10)
+
+    # Array w has shape (3, 10, 10)
+    w = tensor.asarray([x_cpu, x_gpu, x_np], device="level_zero:gpu")
+
+Migration may also occur during calls to other array creation functions, e.g. :py:func:`full` when the `fill_value` parameter is an instance
+of :py:class:`usm_ndarray`. In such a case default values of device placement keywords are interpreted to avoid data migration, i.e. the
+new array is created on the same device where `fill_value` array was allocated.
+
+.. code-block:: python
+    :caption: Example: Using ``usm_ndarray`` as arguments to array construction _dpctl_tensor_utility_functions
+
+    from dpctl import tensor
+
+    # Zero-dimensional array allocated on CPU device
+    pi_on_device = tensor.asarray(tensor.pi, dtype=tensor.float32, device="cpu")
+
+    # x will also be allocated on CPU device
+    x = tensor.full(shape=(100, 100), fill_value=pi_on_device)
+
+    # Create array on GPU. Migration of `pi_on_device` to GPU via host
+    # takes place under the hood
+    y_gpu = tensor.full(shape=(100, 100), fill_value=pi_on_device, device="gpu")
+
+
+Combining arrays with different USM types
+-----------------------------------------
+
+For functions with single argument the returned array has the same ``usm_type`` as the input array.
+
+Functions that combine several ``usm_ndarray`` instances the ``usm_type`` of the output array is determined
+using the following coercion rule:
+
++------------+----------+----------+----------+
+|            | "device" | "shared" | "host"   |
++------------+----------+----------+----------+
+| "device"   | "device" | "device" | "device" |
++------------+----------+----------+----------+
+| "shared"   | "device" | "shared" | "shared" |
++------------+----------+----------+----------+
+| "host"     | "device" | "shared" | "host"   |
++------------+----------+----------+----------+
+
+If assignign USM-type "device" a score of 0, USM-type "shared" a score of 1, and USM-type "host" a score of 2,
+the USM-type of the output array has the smallest score of all its inputs.
+
+.. currentmodule:: dpctl.utils
+
+The convenience function :py:func:`get_coerced_usm_type` is a convenience function to determine the USM-type
+following this convention:
+
+.. code-block:: python
+
+    from dpctl.utils import get_coerced_usm_type
+
+    # r1 has value "device"
+    r1 = get_coerced_usm_type(["device", "shared", "host"])
+
+    # r2 has value "shared"
+    r2 = get_coerced_usm_type(["shared", "shared", "host"])
+
+    # r3 has value "host"
+    r3 = get_coerced_usm_type(["host", "host", "host"])
+
+Sharing data between devices and Python
```

Comment:
It may be worthwhile to add an example demonstrating what happens when migrating a Numpy array of a data type not supported by the destination device.

Can be done later though.